pgmpy/inference/continuous/base.py

```diff
@@ -0,0 +1,365 @@
+    >>> dist_param = np.array([[0.1], [0.9]])
+    >>> grad_logp, logp = GradLogPDFGaussian(dist_param, model).get_gradient_log_pdf()
+    >>> logp
+    -0.4054597701149426
+    >>> grad_logp
+    array([[ 0.90229885],
+           [-0.01149425]])
+    """
+
+    def __init__(self, distribution_param, model):
+        BaseGradLogPDF.__init__(self, distribution_param, model)
+        self.grad_log, self.log_pdf = self._get_gradient_log_pdf()
+
+    def _get_gradient_log_pdf(self):
+        """
+        Method that finds gradient and its log at position
+        """
+        sub_vec = self.distribution_param - self.model.mean
+        grad = - np.dot(self.model.precision_matrix, sub_vec)
+        log_pdf = 0.5 * np.float(np.dot(sub_vec.T, grad))
+
+        return grad, log_pdf
+
+
+class BaseSimulateHamiltonianDynamics(object):
+    """
+    Base class for proposing new values of position and momentum by simulating Hamiltonian Dynamics.
+
+    Classes inheriting this base class can be passed as an argument for
+    simulate_dynamics in inference algorithms.
+
+    Parameters
+    ----------
+    grad_log_pdf : A subclass of pgmpy.inference.continuous.BaseGradLogPDF
+
+    model : An instance of pgmpy.models.Continuous
+        Model for which DiscretizeTime object is initialized
+
+    position : A 1d array like object
+        Vector representing values of parameter position( or X)
+
+    momentum: A 1d array like object
+        Vector representing the proposed value for momentum (velocity)
+
+    stepsize: Float
+        stepsize for the simulating dynamics
+
+    grad_log_position: A 1d array like object, defaults to None
+        Vector representing gradient log at given position
+        If None, then will be calculated
+
+    Examples
+    --------
+    >>> from pgmpy.inference.continuous import BaseSimulateHamiltonianDynamics
+    >>> from pgmpy.models import JointGaussianDistribution
+    >>> from pgmpy.inference.continuous import GradLogPDFGaussian
+    >>> # Class should initalize self.new_position, self.new_momentum and self.new_grad_logp
+    >>> # self.new_grad_logp represents gradient log at new proposed value of position
+    >>> class ModifiedEuler(BaseSimulateHamiltonianDynamics):
+    ...     def __init__(self, grad_log_pdf, model, position, momentum, stepsize, grad_log_position=None):
+    ...         BaseSimulateHamiltonianDynamics.__init__(self, grad_log_pdf, model, position,
+    ...                                                  momentum, stepsize, grad_log_position)
+    ...         self.new_position, self.new_momentum, self.new_grad_logp = self._get_proposed_values()
+    ...     def _get_proposed_values(self):
+    ...         momentum_bar = self.momentum + self.stepsize * self.grad_log_position
+    ...         position_bar = self.position + self.stepsize * momentum_bar
+    ...         return position_bar, momentum_bar, grad_log_position
+    >>> import numpy as np
+    >>> pos = np.array([[1], [2]])
```

Comment:
@khalibartan Why are you accepting 2-D arrays like this everywhere ? I don't think this is intuitive at all. These are all vectors and should be represented by a 1-D array.